protostar/commands/cairo1_commands/fetch_from_scarb.py

```diff
@@ -0,0 +1,89 @@
+from typing import Optional, Dict, List
+from pathlib import Path
+import os
+import json
+import subprocess
+
+import logging
+
+from protostar.protostar_exception import ProtostarException
+
+
+class ScarbMetadataFetchException(ProtostarException):
+    def __init__(self, message: str, details: Optional[str] = None):
+        super().__init__(
+            "Error while trying to fetch Scarb metadata:\n" + message, details
+        )
+
+
+def has_scarb_toml(project_root_path: Path) -> bool:
+    return "Scarb.toml" in os.listdir(project_root_path)
+
+
+def read_scarb_metadata(scarb_toml_path: Path) -> Dict:
+    try:
+        result = subprocess.run(
+            [
+                "scarb",
+                "--json",
+                "--manifest-path",
+                scarb_toml_path,
+                "metadata",
+                "--format-version",
+                "1",
+            ],
+            check=False,  # dont throw exception on fail
+            cwd=scarb_toml_path,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+        )
+    except subprocess.CalledProcessError as ex:
+        raise ScarbMetadataFetchException(str(ex)) from ex
+
+    if result.returncode != 0:
+        raise ScarbMetadataFetchException(
+            "Metadata fetch returned a non-zero exit code."
+        )
+
+    # only the last line of the output contains metadata
+    result = result.stdout.decode("utf-8").strip().split("\n")[-1]
```

Comment:
You shouldn't decode to UTF-8 here. I would suggest sticking with bytes. Reason: broken output from Scarb could yield non-UTF-8 byte sequences in the output; which could happen to be perfectly understandable by JSON parser

Reply:
"Changed in version 3.6: s can now be of type [bytes](https://docs.python.org/3/library/stdtypes.html#bytes) or [bytearray](https://docs.python.org/3/library/stdtypes.html#bytearray). The input encoding should be UTF-8, UTF-16 or UTF-32."

https://docs.python.org/3/library/json.html#basic-usage